src/controllers/dynakube/activegate/reconciler.go

```diff
@@ -2,263 +2,101 @@ package activegate
-	err := r.Get(context.TODO(), client.ObjectKey{Name: desiredSts.Name, Namespace: desiredSts.Namespace}, &sts)
-	if err != nil {
-		return nil, errors.WithStack(err)
-	}
-	return &sts, nil
-}
-
-func (r *Reconciler) createStatefulSetIfNotExists(desiredSts *appsv1.StatefulSet) (bool, error) {
-	_, err := r.getStatefulSet(desiredSts)
-	if err != nil && k8serrors.IsNotFound(errors.Cause(err)) {
-		log.Info("creating new stateful set for " + r.feature)
-		return true, r.Create(context.TODO(), desiredSts)
-	}
-	return false, err
-}
-
-func (r *Reconciler) updateStatefulSetIfOutdated(desiredSts *appsv1.StatefulSet) (bool, error) {
-	currentSts, err := r.getStatefulSet(desiredSts)
-	if err != nil {
+	if err := r.reconcileActiveGateProxySecret(); err != nil {
 		return false, err
 	}
-	if !kubeobjects.HasChanged(currentSts, desiredSts) {
-		return false, nil
-	}
-
-	if kubeobjects.LabelsNotEqual(currentSts.Spec.Selector.MatchLabels, desiredSts.Spec.Selector.MatchLabels) {
-		return r.recreateStatefulSet(currentSts, desiredSts)
-	}
 
-	log.Info("updating existing stateful set")
-	if err = r.Update(context.TODO(), desiredSts); err != nil {
-		return false, err
+	var caps = capability.GenerateActiveGateCapabilities(r.Dynakube)
+	for _, agCapability := range caps {
+		if agCapability.Enabled() {
+			return r.createCapability(agCapability)
+		} else {
+			if err := r.deleteCapability(agCapability); err != nil {
+				return false, err
+			}
+		}
 	}
 	return true, err
 }
 
-func (r *Reconciler) recreateStatefulSet(currentSts, desiredSts *appsv1.StatefulSet) (bool, error) {
-	log.Info("immutable section changed on statefulset, deleting and recreating", "name", desiredSts.Name)
-
-	err := r.Delete(context.TODO(), currentSts)
-	if err != nil {
-		return false, err
+func (r *Reconciler) reconcileActiveGateProxySecret() (err error) {
+	gen := capability.NewActiveGateProxySecretGenerator(r.Client, r.apiReader, r.Dynakube.Namespace, log)
+	if r.Dynakube.NeedsActiveGateProxy() {
+		return gen.GenerateForDynakube(r.context, r.Dynakube)
+	} else {
+		return gen.EnsureDeleted(r.context, r.Dynakube)
 	}
-
-	log.Info("deleted statefulset")
-	log.Info("recreating statefulset", "name", desiredSts.Name)
-
-	return true, r.Create(context.TODO(), desiredSts)
 }
 
-func (r *Reconciler) deleteStatefulSetIfOldLabelsAreUsed(desiredSts *appsv1.StatefulSet) (bool, error) {
-	currentSts, err := r.getStatefulSet(desiredSts)
-	if err != nil {
-		return false, err
-	}
-
-	if !reflect.DeepEqual(currentSts.Labels, desiredSts.Labels) {
-		log.Info("deleting existing stateful set")
-		if err = r.Delete(context.TODO(), desiredSts); err != nil {
-			return false, err
-		}
-		return true, nil
-	}
-
-	return false, nil
+func (r *Reconciler) createCapability(agCapability capability.Capability) (updated bool, err error) {
+	return capability.
+		NewReconciler(r.Client, agCapability, core.NewReconciler(r.Client, r.apiReader, r.scheme, r.Dynakube, agCapability), r.Dynakube).
+		Reconcile()
```

Comment:
introduces dependencies on capability and core reconcilers which cannot be mocked for testing

Reply:
Fixed with dependency injection.